src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -1865,25 +1869,53 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
 
         $container->getDefinition('messenger.retry_strategy_locator')
             ->replaceArgument(0, $transportRetryReferences);
+        
+        $hasFailureTransports = false;
+        $failureTransportsByTransportName = [];
+        $failureTransportsByTransportNameServiceLocatorId = 'messenger.failure_transports_by_transport_name.locator';
 
         if ($config['failure_transport']) {
             if (!isset($senderReferences[$config['failure_transport']])) {
                 throw new LogicException(sprintf('Invalid Messenger configuration: the failure transport "%s" is not a valid transport or service id.', $config['failure_transport']));
             }
 
-            $container->getDefinition('messenger.failure.send_failed_message_to_failure_transport_listener')
-                ->replaceArgument(0, $senderReferences[$config['failure_transport']]);
+            $hasFailureTransports = true;
+            $failureTransports[$config['failure_transport']] = $senderReferences[$config['failure_transport']];
+            $container->setAlias('messenger.failure_transports.default', $config['failure_transport']);
+        }
+
+        foreach ($config['transports'] as $name => $transport) {
+            if ($transport['failure_transport']) {
+                if (!isset($config['transports'][$transport['failure_transport']])) {
+                    throw new LogicException(sprintf('Invalid Messenger configuration: the failure transport "%s" is not a valid transport or service id.', $transport['failure_transport']));
+                }
+
+                $failureTransportsByTransportName[$name] = $senderReferences[$transport['failure_transport']];
+                $hasFailureTransports = true;
+            } else {
+                $failureTransportsByTransportName[$name] = $senderReferences[$config['failure_transport']] ?? null;
+            }
+        }
+
+        if ($hasFailureTransports) {
+            $globalFailureReceiver = $config['failure_transport'] ?? null;
             $container->getDefinition('console.command.messenger_failed_messages_retry')
-                ->replaceArgument(0, $config['failure_transport']);
+                ->replaceArgument(0, $globalFailureReceiver);
             $container->getDefinition('console.command.messenger_failed_messages_show')
-                ->replaceArgument(0, $config['failure_transport']);
+                ->replaceArgument(0, $globalFailureReceiver);
             $container->getDefinition('console.command.messenger_failed_messages_remove')
-                ->replaceArgument(0, $config['failure_transport']);
+                ->replaceArgument(0, $globalFailureReceiver);
+
+            $failureTransportsByTransportNameServiceLocator = ServiceLocatorTagPass::register($container, $failureTransportsByTransportName, $failureTransportsByTransportNameServiceLocatorId);
+            $container->getDefinition($failureTransportsByTransportNameServiceLocatorId)
```

Comment:
I don't *think* that we need this `$failureTransportsByTransportNameServiceLocatorId` variable - most of the time, we don't pass a 3rd argument to `ServiceLocatorTagPass::register()`. But I'm not actually sure what you're needing to accomplish here. Why do we need to replace the first (0) argument on the next line?

Reply:
Related, I'm not sure that we need to manually create a `messenger.failure_transports_by_transport_name.locator` service in the `messenger.php` file. I think (?) usually when we call `ServiceLocatorTagPass::register()`, THAT registers a service locator service for us - it doesn't ned to be already-registered somewhere.

Reply:
This has already been fixed in this [commit](https://github.com/symfony/symfony/pull/38468/commits/19ed68cb764cebd6cd6a19e0fea420eb25c2dfc1)